src/hotspot/cpu/ppc/stubGenerator_ppc.cpp

```diff
@@ -3544,6 +3544,462 @@ class StubGenerator: public StubCodeGenerator {
+
+    // According to the ELF V2 ABI, registers r3-r12 are volatile and available for use without save/restore
+    Register s      = R3_ARG1; // source starting address of Base64 characters
+    Register sp     = R4_ARG2; // source offset
+    Register sl     = R5_ARG3; // source length = # of Base64 characters to be processed
+    Register d      = R6_ARG4; // destination address
+    Register dp     = R7_ARG5; // destination offset
+    Register isURL  = R8_ARG6; // boolean, if non-zero indicates use of RFC 4648 base64url encoding
+
+    // Local variables
+    Register const_ptr     = R9;  // used for loading constants
+    Register tmp_reg       = R10; // used for speeding up load_constant_optimized()
+
+    // Re-use R9 and R10 to avoid using non-volatile registers (requires save/restore)
+    Register out           = R9;  // moving out (destination) pointer
+    Register in            = R10; // moving in (source) pointer
+
+    // Volatile VSRS are 0..13, 32..51 (VR0..VR13)
+    // VR Constants
+    VectorRegister  vec_0s                  = VR0;
+    VectorRegister  vec_4s                  = VR1;
+    VectorRegister  vec_8s                  = VR2;
+    VectorRegister  vec_special_case_char   = VR3;
+    VectorRegister  pack_rshift             = VR4;
+    VectorRegister  pack_lshift             = VR5;
+    // P10+
+    VectorRegister  vec_0x3fs               = VR4; // safe to reuse pack_rshift's register
+
+    // VSR Constants
+    VectorSRegister offsetLUT               = VSR0;
+    VectorSRegister maskLUT                 = VSR1;
+    VectorSRegister bitposLUT               = VSR2;
+    VectorSRegister vec_0xfs                = VSR3;
+    VectorSRegister vec_special_case_offset = VSR4;
+    VectorSRegister pack_permute            = VSR5;
+
+    // Variables for lookup
+    // VR
+    VectorRegister  input                   = VR6;
+    VectorRegister  higher_nibble           = VR7;
+    VectorRegister  eq_special_case_char    = VR8;
+    VectorRegister  offsets                 = VR9;
+    VectorRegister  non_match               = VR10;
+
+    // VSR
+    VectorSRegister bit                     = VSR6;
+    VectorSRegister lower_nibble            = VSR7;
+    VectorSRegister M                       = VSR8;
+
+    // Variables for pack
+    // VR
+    VectorRegister  l                       = VR7;  // reuse higher_nibble's register
+    VectorRegister  r                       = VR8;  // reuse eq_special_case_char's register
+    VectorRegister  gathered                = VR9;  // reuse offsets's register
+
+    Label not_URL, calculate_size, unrolled_loop_start, skip_xxsel[loop_unrolls], unrolled_loop_exit;
+
+    // The upper 32 bits of the non-pointer parameter registers are not
+    // guaranteed to be zero, so mask off those upper bits.
+    __ clrldi(sp, sp, 32);
+    __ clrldi(sl, sl, 32);
+    __ clrldi(dp, dp, 32);
+    __ clrldi(isURL, isURL, 32);
+
+    // Load constant vec registers that need to be loaded from memory
```

Comment:
With larger unroll factor we run through this code more often without making any progress, because only the Java part does all the work for the remaining bytes. Would be nice to move unnecessary parts for that between mtctr and align.

Reply:
You're right that there's quite a lot of set up before the size check is performed.  I will fix, this and run regression the regression test.